File: src/lib/store/auto-reply.ts
// 自动回复 store：规则（按编辑账户）+ 建议回复队列（聚合）。
// dismiss 乐观移除 + 失败回滚，仿 mail store setFlagOptimistic。
// 「去回复」不在本 store（跨 compose/ui store），由队列组件接线。

import { create } from 'zustand';

import * as tauri from '../tauri';
import type { AutoReplyRule, AutoReplyRuleInput, SuggestedReply } from '../types';
import { errMsg } from '../utils';

interface AutoReplyState {
  rules: AutoReplyRule[];
  rulesAccountId: string | null;
  queue: SuggestedReply[];
  loadingRules: boolean;
  loadingQueue: boolean;
  error: string | null;

  loadRules: (accountId: string) => Promise<void>;
  addRule: (input: AutoReplyRuleInput) => Promise<void>;
  updateRule: (rule: AutoReplyRule) => Promise<void>;
  removeRule: (id: string) => Promise<void>;
  toggleRule: (id: string, enabled: boolean) => Promise<void>;

  loadQueue: () => Promise<void>;
  dismiss: (id: string) => Promise<void>;
  clearError: () => void;
}

export const useAutoReplyStore = create<AutoReplyState>((set, get) => ({
  rules: [],
  rulesAccountId: null,
  queue: [],
  loadingRules: false,
  loadingQueue: false,
  error: null,

  loadRules: async (accountId) => {
    set({ loadingRules: true, rulesAccountId: accountId, error: null });
    try {
      const rules = await tauri.autoReplyRulesList(accountId);
      if (get().rulesAccountId === accountId) set({ rules });
    } catch (e) {
      set({ error: errMsg(e) });
    } finally {
      if (get().rulesAccountId === accountId) set({ loadingRules: false });
    }
  },

  addRule: async (input) => {
    set({ error: null });
    try {
      await tauri.autoReplyRuleAdd(input);
      await get().loadRules(input.accountId);
    } catch (e) {
      set({ error: errMsg(e) });
    }
  },

  updateRule: async (rule) => {
    set({ error: null });
    try {
      await tauri.autoReplyRuleUpdate(rule);
      await get().loadRules(rule.accountId);
    } catch (e) {
      set({ error: errMsg(e) });
    }
  },

  removeRule: async (id) => {
    const accountId = get().rulesAccountId;
    set({ error: null });
    try {
      await tauri.autoReplyRuleRemove(id);
      if (accountId !== null) await get().loadRules(accountId);
    } catch (e) {
      set({ error: errMsg(e) });
    }
  },

  toggleRule: async (id, enabled) => {
    const accountId = get().rulesAccountId;
    set({ error: null });
    try {
      await tauri.autoReplyRuleSetEnabled(id, enabled);
      if (accountId !== null) await get().loadRules(accountId);
    } catch (e) {
      set({ error: errMsg(e) });
    }
  },

  loadQueue: async () => {
    set({ loadingQueue: true });
    try {
      const queue = await tauri.suggestedRepliesList();
      set({ queue, error: null });
    } catch (e) {
      set({ error: errMsg(e) });
    } finally {
      set({ loadingQueue: false });
    }
  },

  dismiss: async (id) => {
    // 乐观移除并清旧错误。失败时不还原陈旧整列快照（会复活并发 dismiss 的其它已移除项），
    // 改为从后端重拉权威队列、再写 error（loadQueue 成功路径会清 error，故 error 后置），
    // 仿 mail.deleteMessage 的失败重载惯例。
    set({ queue: get().queue.filter((s) => s.id !== id), error: null });
    try {
      await tauri.suggestedReplyDismiss(id);
    } catch (e) {
      await get().loadQueue();
      set({ error: errMsg(e) });
    }
  },

  clearError: () => {
    set({ error: null });
  },
}));

File: src-tauri/src/commands/auto_reply.rs
//! 自动回复规则 CRUD + 建议回复队列命令。设置 UI / 自动回复中心是唯一调用方。
//! 本文件不发送邮件——发送恒由前端 compose tab 经 smtp_send 人工触发。

use tauri::State;
use uuid::Uuid;

use crate::db::auto_reply_rules::{self, AutoReplyRule, AutoReplyRuleInput};
use crate::db::suggested_replies::{self, SuggestedReply};
use crate::error::{AppError, AppResult};
use crate::AppState;

const CATEGORIES: [&str; 5] = ["personal", "work", "notification", "promotion", "spam"];

/// 校验：name/draft_intent 不得为空白；category 必须在闭集；priority_ceiling 必须 1..=3。
fn validate(
    name: &str,
    draft_intent: &str,
    category: &Option<String>,
    ceiling: Option<i64>,
) -> AppResult<()> {
    if name.trim().is_empty() {
        return Err(AppError::Config("rule name must not be empty".into()));
    }
    if draft_intent.trim().is_empty() {
        return Err(AppError::Config("draft intent must not be empty".into()));
    }
    if let Some(c) = category {
        if !CATEGORIES.contains(&c.as_str()) {
            return Err(AppError::Config(format!("unknown category: {c}")));
        }
    }
    if let Some(p) = ceiling {
        if !(1..=3).contains(&p) {
            return Err(AppError::Config(format!(
                "priority ceiling must be 1..=3, got {p}"
            )));
        }
    }
    Ok(())
}

#[tauri::command]
pub async fn auto_reply_rules_list(
    state: State<'_, AppState>,
    account_id: Uuid,
) -> AppResult<Vec<AutoReplyRule>> {
    auto_reply_rules::list_by_account(state.pool().await?, account_id).await
}

#[tauri::command]
pub async fn auto_reply_rule_add(
    state: State<'_, AppState>,
    input: AutoReplyRuleInput,
) -> AppResult<AutoReplyRule> {
    validate(
        &input.name,
        &input.draft_intent,
        &input.match_category,
        input.match_priority_ceiling,
    )?;
    auto_reply_rules::insert(state.pool().await?, &input).await
}

#[tauri::command]
pub async fn auto_reply_rule_update(
    state: State<'_, AppState>,
    rule: AutoReplyRule,
) -> AppResult<()> {
    validate(
        &rule.name,
        &rule.draft_intent,
        &rule.match_category,
        rule.match_priority_ceiling,
    )?;
    auto_reply_rules::update(state.pool().await?, &rule).await
}

#[tauri::command]
pub async fn auto_reply_rule_remove(state: State<'_, AppState>, id: Uuid) -> AppResult<()> {
    auto_reply_rules::delete(state.pool().await?, id).await
}

#[tauri::command]
pub async fn auto_reply_rule_set_enabled(
    state: State<'_, AppState>,
    id: Uuid,
    enabled: bool,
) -> AppResult<()> {
    auto_reply_rules::set_enabled(state.pool().await?, id, enabled).await
}

#[tauri::command]
pub async fn suggested_replies_list(state: State<'_, AppState>) -> AppResult<Vec<SuggestedReply>> {
    suggested_replies::list_pending(state.pool().await?).await
}

#[tauri::command]
pub async fn suggested_reply_dismiss(state: State<'_, AppState>, id: Uuid) -> AppResult<()> {
    suggested_replies::dismiss(state.pool().await?, id).await
}

#[cfg(test)]
mod tests {
    use super::validate;

    #[test]
    fn validate_accepts_valid() {
        assert!(validate("工作", "确认", &None, None).is_ok());
        assert!(validate("r", "i", &Some("work".into()), Some(1)).is_ok());
        assert!(validate("r", "i", &Some("spam".into()), Some(3)).is_ok());
    }
    #[test]
    fn validate_rejects_bad_category_and_ceiling() {
        assert!(validate("r", "i", &Some("nope".into()), None).is_err());
        assert!(validate("r", "i", &None, Some(0)).is_err());
        assert!(validate("r", "i", &None, Some(4)).is_err());
    }
    #[test]
    fn validate_rejects_empty_name_or_intent() {
        assert!(validate("", "i", &None, None).is_err());
        assert!(validate("   ", "i", &None, None).is_err());
        assert!(validate("r", "", &None, None).is_err());
        assert!(validate("r", "  ", &None, None).is_err());
    }
}
